source/debian/spyview3d.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SPYVIEW SECTION "May  6, 2003"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
spyview \- program to do something
.SH SYNOPSIS
.B spyview3d
.RI " files" ...
.br
.SH DESCRIPTION
.PP 
A 3d version of spyview implemented with OpenGL.
.SH GUI SHORTCUTS
.PP
.B Button 1
is used for rotating the 3D surface. A rotation of a rigid body is
defined by three Euler angles: Theta, the rotation of the body Z-axis
away from the original Z-axis, Psi, the rotation of the body about
it's Z-axis, and Phi, the azimuthal rotation of the body Z-axis about
the original Z-axis.
.PP
In spyview3d, Phi is fixed so that the image is always tilted towards
you. The vertical motion of the mouse will adjust theta, the angle of
this tilt, and the horizontal motion of the mouse will adjust the
rotation of the body about it's z axis.
.PP
.B Button 2
will adjust the scaling of the image. Horizontal motion will scale all
three dimensions (zooming), while vertical motion will only adjust the
z scaling of the surface.
.PP
.B Button 3
allows you to translate the image.
.PP
.B c
Show/hide the controls window.
.SH CONTROL WINDOW SETTINGS
.PP
.B Grey
will draw the surface all in one color. This is useful for
investigating the effects of lighting.
.PP 
.B Colormap settings
These give you some simple controls over how the colormap is applied to the
image.
.PP
.B Save
Render the opengl image to the specified format and filename using
gl2ps: http://www.geuz.org/gl2ps/

.SH LIGHTING CONTROLS
.PP
.B Type
The light source can be set to be a point source or be directional.
In the case of a directional source, the direction is set by the phi
and theta angles, and the radius slider changes only the position of
the light position indicator.
.PP
.B Ambient
Adjust the intensity of the ambient lighting: this is a light source
that lights the the surface uniformly from all directions.
.PP
.B Diffuse
Adjust the intensity of the diffuse lighting: this light is radiated
from a point source at the position given by the sliders below.  (In
directional mode, the light is instead a plane was, as if from a point
source at infinity.) With ambient lighting, light that hits the
surface is re-radiated in all directions.
.PP
.B Specular
Adjust the intensity of the specular lighting. Specular lighting
reflects off of the surface (instead of being re-radiated). (Haven't
got it to work yet...)
.PP
For more info about OpenGL lighting, see
.PP
http://www.sjbaker.org/steve/omniv/opengl_lighting.html.
.SH SEE ALSO
.BR spyview(1),
.BR spybrowse(1)
.SH AUTHOR
This manual page was written by Gary Steele <gsteele@electron.mit.edu>,
for the Debian project (but may be used by others).


debian/changelog
spyview (0.1-38) local; urgency=low

  * Serious changes since the last update!

 -- gsteele <gsteele@newkavli.nano.tudelft.nl>  Sat, 18 Aug 2007 11:02:30 -0400

spyview (0.1-37) local; urgency=low

  * Added loads of stuff, including the sweet new postscript output, and 
    an updated man page.

 --  <gsteele@qtmartini.sqt.tn.tudelft.nl>  Fri, 24 Nov 2006 16:43:46 +0100

spyview (0.1-36) local; urgency=low

  * Minor update to man page, very significant update to the software...

 --  <gsteele@qtmartini.sqt.tn.tudelft.nl>  Wed, 25 Oct 2006 09:55:37 +0200

spyview (0.1-35) local; urgency=low

  * Too many new features to list...including severely outdated man pages...

 -- odie <gsteele@qtmartini.sqt.tn.tudelft.nl>  Fri, 25 Aug 2006 11:58:50 +0200

spyview (0.1-34) local; urgency=low

  * Added pdf files and a lot of new features.

 -- odie <odie@electron.mit.edu>  Fri, 19 May 2006 05:43:36 -0400

spyview (0.1-33) local; urgency=low

  * Included stminfo in the distribution.

 --  <gsteele@vortex.ashoori.mit.edu>  Mon, 25 Apr 2005 16:27:49 -0400

spyview (0.1-32) local; urgency=low

  * Added support for a 3D binary floating point format that I call "mtx"
  * Added HSV rotation wheel for adjusting colors of a colormap
  * Added a new blue-green-pink colormap designed specifically for 
    my incompressible strip simulation data, as well as a new transfer 
    function adjustment that centers the colormap on the peak of the 
    histogram
  * Added a column-by-column subtraction routine to the image processing.

 --  <gsteele@vortex.ashoori.mit.edu>  Sat, 22 Jan 2005 15:53:15 -0500

spyview (0.1-31) local; urgency=low

  * UI enhancements, added support for arbitrary sized colormaps
  * Parameters are now saved in PPM file as comments
  * Modified matlabpink colormap so that it does not have a pre-applied 
    gamma correction

 --  <gsteele@vortex.ashoori.mit.edu>  Sun, 21 Nov 2004 17:47:10 -0500

spyview (0.1-30) local; urgency=low

  * Improved histogram plot

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Nov 2004 10:36:04 -0500

spyview (0.1-29) local; urgency=low

  * Big enhancements to spybrowse: added a new Fl_Listbox widget

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Nov 2004 08:49:57 -0500

spyview (0.1-28) local; urgency=low

  * Added support for separate x and y zooms, updated man page.

 --  <gsteele@vortex.ashoori.mit.edu>  Thu,  5 Aug 2004 16:45:20 -0400

spyview (0.1-27) local; urgency=low

  * Increased line buffer size of dat2pgm to 1024*1024 bytes. Also 
    fixed the logarithm so that it correctly output the base10 exponent.

 --  <gsteele@vortex.ashoori.mit.edu>  Wed,  4 Aug 2004 16:35:58 -0400

spyview (0.1-26) local; urgency=low

  * Removed ugly colormaps again after they snuck back in.

 --  <gsteele@vortex.ashoori.mit.edu>  Tue, 27 Jul 2004 12:52:10 -0400

spyview (0.1-25) local; urgency=low

  * Added a program dat2pgm for converting matrix ascii files into 
    pgm files that include the spyview comment fields for the axis
    scaling.

 --  <gsteele@vortex.ashoori.mit.edu>  Wed, 21 Jul 2004 21:09:09 -0400

spyview (0.1-24) local; urgency=low

  * Added a bunch of stuff, including arbitrary line cuts, support for
    changing x,y,z scaling/units of image using a dialog box, and other
    great stuff. Also, took topometrix colormaps out of distribution, since 
    they were just clogging things up and were mostly crap anyway.

 --  <gsteele@vortex.ashoori.mit.edu>  Fri,  2 Apr 2004 20:23:13 -0500

spyview (0.1-23) local; urgency=low

  * Added more normalize on load options to spyview.

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Mar 2004 14:54:15 -0500

spyview (0.1-22) local; urgency=low

  * Added some colormaps, finally fixed filename stuff so we can load
    files that are in subdirectories.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 20 Jan 2004 18:30:16 -0500

spyview (0.1-21) local; urgency=low

  * Fixed lbl subtraction.

 -- Gary Steele <gsteele@electron.mit.edu>  Sat, 10 Jan 2004 19:45:21 -0500

spyview (0.1-20) local; urgency=low

  * Added a new program spyrotate 
  * Some code cleanup

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Tue,  9 Dec 2003 14:37:12 -0500

spyview (0.1-19) local; urgency=low

  * Some bugfixes, changes shortcut 's' to save file as ppm. 

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Wed, 26 Nov 2003 10:50:48 -0500

spyview (0.1-18) local; urgency=low

  * Fixed a bug with the plane subtraction in the saveFile routine.

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Sun, 19 Oct 2003 13:33:22 -0400

spyview (0.1-17) local; urgency=low

  * Added an shlib.local file to override libc dependecy.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue,  2 Sep 2003 15:15:50 -0400

spyview (0.1-16) local; urgency=low

  * Recompiled with gcc-3.3

 -- Gary Steele <gsteele@electron.mit.edu>  Tue,  2 Sep 2003 12:38:16 -0400

spyview (0.1-15) local; urgency=low

  * Fixed up gl2ps scaling, added rollers for colormap in spyview3d

 -- Gary Steele <gsteele@electron.mit.edu>  Sun,  3 Aug 2003 17:04:58 -0400

spyview (0.1-14) local; urgency=low

  * Changed gnuplot xsection commands so you can zoom with the mouse
    properly.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 29 Jul 2003 15:09:35 -0400

spyview (0.1-13) local; urgency=low

  * Added plane subtraction to spyview3d. 

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 22:29:31 -0400

spyview (0.1-12) local; urgency=low

  * Added basic colormap manipulation support to spyview3d.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 16:16:21 -0400

spyview (0.1-11) local; urgency=low

  * Added support for reading unit and scales out of AFM and Stm 
    files. The crosssections are now plots showing physical units.
    (The zaxis unit is ignored for Stm files, since it's usually
    still +/- 10V).
    
 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 14:15:09 -0400

spyview (0.1-10) local; urgency=low


source/spyview/spyview3d.C
#include <stdio.h>
#include "spyview3d.H"
#include "ImageWindow3d.H"
#include "spyview3d_ui.h"

#include <FL/fl_ask.H>
#include <string.h>
#include <FL/Fl_File_Chooser.H>
#include <FL/filename.H>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
// STL stuff
//# include <string.h>
//#include <string>
//#include <vector>

#include "mypam.h"

using namespace std;

int *data;
int *databuf;
int w,h;
vector<string> filenames;
vector<string> cmapfiles;

int global_shortcuts(int event);

int main(int argc, char **argv)
{
  if (argc == 1)
    {
      fprintf(stderr, "Please specify filename\n");
      exit(-1);
    }

  Fl::visual(FL_RGB);
  make_window();
  
  // Construct a list of valid filenames that were given on the input

  int num;
  int fd;
  string fn;
  for (int i=1; i < argc; i++)
    {
      if ( (fd = open(argv[i], O_RDONLY)) != -1)
	{
	  // Make a vector of filenames using the STL
	  fn = argv[i];
	  filenames.push_back(fn);
	  num = filech->add(" ", 0, filech_cb);
	  filech->replace(num, argv[i]);
	  //fprintf(stderr, "adding %s as %s\n", argv[i], (filenames[i-1]).c_str());
	  close(fd);
	}
    }

  if (filech->size() == 0)
    {
      fprintf(stderr, "No valid files found\n");
      exit(-1);
    }

  // Construct a list of colormap files
  // First scan them from /usr/share/spyview/cmaps, then look in ~/cmaps/

  int n;
  struct dirent **namelist;

  string share_path = "/usr/share/spyview/cmaps";
  n = fl_filename_list(share_path.c_str(), &namelist, fl_casealphasort);
  for (int i = 0; i<n ; i++)
    {
      if (strstr(namelist[i]->d_name, ".ppm") != NULL)
	
	{
	  cmapch->add(namelist[i]->d_name, 0, 0);
	  fn = namelist[i]->d_name;
	  cmapfiles.push_back(share_path + "/" + fn);
	}
    }

  int m;
  char *home = getenv("HOME"); 
  string home_path = home;
  home_path += "/cmaps";
  n = fl_filename_list(home_path.c_str(), &namelist, fl_casealphasort);
  for (int i = 0; i<n ; i++)
    {
      if (strstr(namelist[i]->d_name, ".ppm") != NULL)
	
	{
	  fn = namelist[i]->d_name;	  
	  m = cmapch->add(("~ "+fn).c_str(), 0, 0);
	  cmapfiles.push_back(home_path + "/" + fn);
	}
    }

  options->add("fft");
  options->add("ac");
  options->add("log");
  options->add("sub lbl");

  iw3d->external_update = update_gui;
  update_gui();
  imagewindow->show();
  iw3d->show();
  control->show();

  // Load the file
  filech->value(0);
  filech->do_callback(filech, (void *)argv[1]);
  cmapch->value(17); //mountain-lakes colormap
  cmapch->do_callback();
  Fl::add_handler(&global_shortcuts);
  Fl::run();
}

int global_shortcuts(int event)
{
  int n;
  switch (event) 
    {
    case FL_SHORTCUT:
      n = Fl::event_key();
      switch (n) 
	{
	case 'c':
	  if (control->shown())
	    control->hide();
	  else
	    control->show();
	  return 1;
	  break;
	case 'q':
	  exit(0);
	case FL_Right:
	case ' ':
	case 'f':
	  n = filech->value();
	  if (n == filech->size() - 2)
	    {
	      filech->value(0);
	    }
	  else
	    filech->value(n+1);
	  filech->do_callback();
	  return 1;
	  break;
	case FL_Left:
	case FL_BackSpace:
	case 'd':
	  n = filech->value();
	  if (n == 0)
	    filech->value(filech->size()-2);
	  else
	    filech->value(n-1);
	  filech->do_callback();
	  return 1;
	  break;
	case FL_Down:
	case 'j':
	  n = cmapch->value();
	  if (n == cmapch->size() - 2)
	    cmapch->value(0);
	  else
	    cmapch->value(n+1);
	  cmapch->do_callback();
	  return 1;
	  break;
	case FL_Up:
	case 'k':
	  n = cmapch->value();
	  if (n == 0)
	    cmapch->value(cmapch->size()-2);
	  else
	    cmapch->value(n-1);
	  cmapch->do_callback();
	  return 1;
	  break;
	}
    }
  return 0;
}

void filech_cb(Fl_Widget *o, void *)
{
  const char *filename = filenames[filech->value()].c_str();
  iw3d->loadData(filename);
  char label[1024];
  snprintf(label, 1024, "%s", filech->text());
  imagewindow->label(label);
  char buf[1024];
  char *p;
  strncpy(buf, filename, 1024);
  p = strrchr(buf, '.');
  if (p == NULL)
    p = buf + strlen(buf);
  strcpy(p, ".out");
  savebox->value(buf);
  iw3d->redraw();
}

void saveb_cb(Fl_Widget *, void *)
{
  GLint format;
  if (postscriptbutton->value()) 
    format = GL2PS_PS;
  else if (epsbutton->value()) 
    format = GL2PS_EPS;
  else if (latexbutton->value())
    format = GL2PS_TEX;
  iw3d->saveFile(savebox->value(), format);
  fl_message("File was saved to %s", savebox->value());
}


void cmapch_cb(Fl_Widget *o, void*)
{
  FILE *fp;
  const char *filename = cmapfiles[cmapch->value()].c_str();
  fp = fopen(filename, "r");
  if (fp == NULL)
    {
      perror(filename);
      exit(-1);
    }
  
  int rows, cols;
  pixel **image;
  pixval maxval;

  image = ppm_readppm(fp, &cols, &rows, &maxval);
  fclose(fp);

  if (cols > 1)
    {
      fprintf(stderr, "Invalid colormap %s: must contain only one column!\n", filename);
      exit(-1);
    }

  if (rows > 256)
    {
      fprintf(stderr, "Invalid colormap %s: must contain 256 rows!\n", filename);
      exit(-1);
    }
  
  if (maxval != 255)
    {
      fprintf(stderr, "Invalid colormap %s: color depth must be 24 bit (255 maxval)\n", filename);
      exit(-1);
    }

  for (int i=0; i<rows; i++)
    {
      iw3d->colormap[0][i] = image[0][i].r;
      iw3d->colormap[1][i] = image[0][i].g;
      iw3d->colormap[2][i] = image[0][i].b;
    }

  ppm_freearray(image, rows);
  iw3d->redraw();
  char label[1024];
  snprintf(label, 1024, "%s - %s", filech->text(), cmapch->text());
  iw3d->label(label);
}

void update_gui()
{
  psival->value(iw3d->getPsi());
  thetaval->value(iw3d->getTheta());
  zoomval->value(iw3d->getScaleX());
  zscaleval->value(iw3d->getScaleZ()/iw3d->getScaleX());
  xtranval->value(iw3d->getTranslateX());
  ytranval->value(iw3d->getTranslateY());
  //fprintf(stderr, "% 6.2f % 6.2f % 6.2f\n", iw3d->getScaleX(), iw3d->getScaleY(), iw3d->getScaleZ());
}
